### ISSUE TITLE
Extract date and time from video files

### ISSUE THREAD

**Issue by Quintasan** [ayoisaiah/f2#28]
Hi,

I was trying to sort out my media files, and I've noticed that f2 doesn't really support reading date and time reading from a mp4 file. They don't have the usual date time field used for photos but have `MediaCreateDate`:

```json
[{
  "SourceFile": "testdata/videos/pixel6pro.mp4",
  "ExifToolVersion": 11.88,
  "FileName": "pixel6pro.mp4",
  "Directory": "testdata/videos",
  "FileSize": "15 MB",
  "FileModifyDate": "2022:10:24 03:46:40+02:00",
  "FileAccessDate": "2022:10:24 17:24:54+02:00",
  "FileInodeChangeDate": "2022:10:24 17:50:10+02:00",
  "FilePermissions": "rwxrwxrwx",
  "FileType": "MP4",
  "FileTypeExtension": "mp4",
  "MIMEType": "video/mp4",
  "MajorBrand": "MP4  Base Media v1 [IS0 14496-12:2003]",
  "MinorVersion": "2.0.0",
  "CompatibleBrands": ["isom","iso2","mp41"],
  "MediaDataSize": 15765030,
  "MediaDataOffset": 44,
  "MovieHeaderVersion": 0,
  "CreateDate": "2022:10:21 15:58:43",
  "ModifyDate": "2022:10:21 15:58:43",
  "TimeScale": 10000,
  "Duration": "8.01 s",
  "PreferredRate": 1,
  "PreferredVolume": "100.00%",
  "PreviewTime": "0 s",
  "PreviewDuration": "0 s",
  "PosterTime": "0 s",
  "SelectionTime": "0 s",
  "SelectionDuration": "0 s",
  "CurrentTime": "0 s",
  "NextTrackID": 6,
  "GPSCoordinates": "40 deg 42' 43.56\" N, 74 deg 0' 42.48\" W",
  "TrackHeaderVersion": 0,
  "TrackCreateDate": "2022:10:21 15:58:43",
  "TrackModifyDate": "2022:10:21 15:58:43",
  "TrackID": 1,
  "TrackDuration": "8.01 s",
  "TrackLayer": 0,
  "TrackVolume": "100.00%",
  "ImageWidth": 1024,
  "ImageHeight": 768,
  "GraphicsMode": "srcCopy",
  "OpColor": "0 0 0",
  "CompressorID": "avc1",
  "SourceImageWidth": 1024,
  "SourceImageHeight": 768,
  "XResolution": 72,
  "YResolution": 72,
  "BitDepth": 24,
  "PixelAspectRatio": "65536:65536",
  "ColorRepresentation": "nclx 1 1 1",
  "VideoFrameRate": 24.233,
  "Warning": "[minor] The ExtractEmbedded option may find more tags in the media data",
  "Balance": 0,
  "AudioFormat": "mp4a",
  "AudioChannels": 2,
  "AudioBitsPerSample": 16,
  "AudioSampleRate": 48000,
  "MatrixStructure": "0 1 0 -1 0 0 0 0 1",
  "MediaHeaderVersion": 0,
  "MediaCreateDate": "2022:10:21 15:58:43",
  "MediaModifyDate": "2022:10:21 15:58:43",
  "MediaTimeScale": 90000,
  "MediaDuration": "0 s",
  "HandlerType": "NRT Metadata",
  "HandlerDescription": "MetaHandle",
  "MetaFormat": "mett",
  "ImageSize": "1024x768",
  "Megapixels": 0.786,
  "AvgBitrate": "15.7 Mbps",
  "GPSLatitude": "40 deg 42' 43.56\" N",
  "GPSLongitude": "74 deg 0' 42.48\" W",
  "Rotation": 90,
  "GPSPosition": "40 deg 42' 43.56\" N, 74 deg 0' 42.48\" W"
}]
```
so technically it should be able to extract this value, parse it and make it accessible via `x.dt.YYYY` for example. What do you think?

I could use `xt.MediaCreateDate` but it's hard to parse it to something like `2022-10-24_18-33` because it uses `:` inside of the string

**Comment by ayoisaiah** [comment]
@Quintasan Thanks for raising this issue.

I think this is a valid use case that will be supported in the next version of F2. I will add the ability to parse any arbitrary string as a date so you can do something like:

```
xt.MediaCreateDate.dt.YYYY
$1.dt.MM
```
...and so on.

If the string is not a valid date (say `xt.ImageSize.dt.YYY`), it will be replaced as is.

**Comment by Quintasan** [comment]
That would solve the problem I have. Thanks!

**Comment by ayoisaiah** [comment]
5f0d95a59ee89b26d7d08de6f7bc6dc4652a2015

**Comment by d-Rickyy-b** [comment]
Hi @ayoisaiah, sorry for necro but I am currently running into the same issue. I would love the name to contain the exif "CreationDate". But when using f2 v1.9.1 it won't rename the file properly.

`exiftool -j IMG_1783.mov`

```json
[
  {
    "SourceFile": "IMG_1783.mov",
    "ExifToolVersion": 12.57,
    "FileName": "IMG_1783.mov",
    "Directory": ".",
    "FileSize": "50 MB",
    "FileModifyDate": "2024:08:12 18:30:19+02:00",
    "FileAccessDate": "2024:08:19 18:40:35+02:00",
    "FileInodeChangeDate": "2024:08:19 01:20:17+02:00",
    "FilePermissions": "-rw-r--r--",
    "FileType": "MOV",
    "FileTypeExtension": "mov",
    "MIMEType": "video/quicktime",
    "MajorBrand": "Apple QuickTime (.MOV/QT)",
    "MinorVersion": "0.0.0",
    "CompatibleBrands": [
      "qt  "
    ],
    "MovieHeaderVersion": 0,
    "CreateDate": "2024:08:12 16:29:15",
    "ModifyDate": "2024:08:12 16:29:20",
    ...
    "TrackCreateDate": "2024:08:12 16:29:15",
    "TrackModifyDate": "2024:08:12 16:29:20",
    "TrackID": 1,
    ...
    "Warning": "[minor] The ExtractEmbedded option may find more tags in the media data",
    "MatrixStructure": "1 0 0 0 1 0 0 0 1",
    "ContentDescribes": "Track 2",
    "MediaHeaderVersion": 0,
    "MediaCreateDate": "2024:08:12 16:29:15",
    "MediaModifyDate": "2024:08:12 16:29:21",
    ...
    "CreationDate": "2024:08:10 20:00:32+02:00",
    ...
  }
]
```

I removed some content from the exiftool output to keep this comment short.

I can't access arbitrary exiftool variables.
```
f2 -f ".+\.mov$" -r 'VID_{x.CreationDate.dt.YYYY}{ext}' -F
*——————————————*——————————————————————————————————————*————————*
|   ORIGINAL   |               RENAMED                | STATUS |
*——————————————*——————————————————————————————————————*————————*
| IMG_1783.mov | VID_{x.CreationDate.dt.YYYY}.mov     | ok     |
| IMG_1880.mov | VID_{x.CreationDate.dt.YYYY} (2).mov | ok     |
| IMG_2933.mov | VID_{x.CreationDate.dt.YYYY} (3).mov | ok     |
| IMG_2934.mov | VID_{x.CreationDate.dt.YYYY} (4).mov | ok     |
*——————————————*——————————————————————————————————————*————————*
 DRY RUN  Commit the above changes with the -x/--exec flag
```

Also cdt seems to return nothing.
```
f2 -f "IMG_\d+\.mov$" -r 'VID_{x.cdt.YYYY}{ext}' -F
*——————————————*——————————————*————————*
|   ORIGINAL   |   RENAMED    | STATUS |
*——————————————*——————————————*————————*
| IMG_1783.mov | VID_.mov     | ok     |
| IMG_1880.mov | VID_ (2).mov | ok     |
| IMG_2933.mov | VID_ (3).mov | ok     |
| IMG_2934.mov | VID_ (4).mov | ok     |
*——————————————*——————————————*————————*
 DRY RUN  Commit the above changes with the -x/--exec flag
```

Do you know why that is? Did I do something wrong?

**Comment by ayoisaiah** [comment]
@d-Rickyy-b It's `xt` not `x`, as in: `xt.CreationDate.dt.YYYY}`

See: https://github.com/ayoisaiah/f2/wiki/Built-in-variables#6-exiftool-variables

**Comment by d-Rickyy-b** [comment]
@ayoisaiah ohh, thank you! It was a bit confusing because I did not realize that using `x` for the provided exif variables and `xt` for exiftool were different things.

Maybe there could be a hint in the wiki about this? Also maybe you can add a link to the section [Exif variables](https://github.com/ayoisaiah/f2/wiki/Built-in-variables#3-exif-variables) that mentions the ability to extract all exiftool fields too?

Anyway, thanks yet again for your help :)

**Comment by ayoisaiah** [comment]
@d-Rickyy-b That's a great point, added the link!
